### PR TITLE
feat(jit): plan ahead auth requirements when required

### PR DESCRIPTION
**Pull request by karatakis**
**Summary:**  
We plan the authentication requirements for the query plan, and we can execute them before we make any calls to remove services.

**Issue Reference(s):**  
Fixes https://github.com/tailcallhq/tailcall/issues/3125

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [x] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

File: src/core/jit/exec_const.rs [hunk 1]
```diff
@@ -52,6 +53,16 @@ impl ConstValueExecutor {
                 .with_errors(vec![Positioned::new(Error::Unknown, Pos::default())]);
         };
 
+        // Attempt to plan authentication
+        let Ok(plan) = AuthPlaner::new(&req_ctx.server.auth)
+            .transform(plan)
+            .to_result()
```
Comment (tusharmath) [inline src/core/jit/exec_const.rs:59]: This is technically not planning. We are updating the plan for each request.

**Review by tusharmath — CHANGES_REQUESTED** [review]

File: src/core/jit/exec_const.rs (hunk 1 above)
Comment (karatakis) [reply src/core/jit/exec_const.rs]: I got it moved to the static query planning generation

File: src/core/jit/exec_const.rs [hunk 2]
```diff
@@ -75,6 +75,13 @@ impl ConstValueExecutor {
         let store = exe.store().await;
         let synth = Synth::new(&plan, store, vars);
 
+        if let Some(_auth_n) = &plan.auth_n {
+            if let Err(err) = req_ctx.auth_ctx.validate(req_ctx).await.to_result() {
+                return Response::default()
+                    .with_errors(vec![Positioned::new(Error::from(err), Pos::default())]);
+            }
+        }
```
Comment (laststylebender14) [inline src/core/jit/exec_const.rs:83]: in ``let store = exe.store().await;` we execute the actual plan, so we would like to do the auth before that

Comment (karatakis) [reply src/core/jit/exec_const.rs:83]: thanks

File: src/core/jit/request.rs [hunk 3]
```diff
@@ -46,8 +46,8 @@ impl Request<ConstValue> {
         let plan = builder.build(self.operation_name.as_deref())?;
 
         transform::CheckConst::new()
+            .pipe(transform::AuthPlaner::new(&blueprint.server.auth))
```
Comment (laststylebender14) [inline src/core/jit/request.rs:49]: AuthPlanner goes through fields and if there's protected IR it set's that in query plan, right?
so we need to run this transformer only when `blueprint.server.auth.is_some()`.

we can utilise the when operator here to do that.

**Review by laststylebender14 — APPROVED** [review]

**Comment by karatakis** [comment]
Closed in favor of https://github.com/tailcallhq/tailcall/pull/3139

**Pull request closed without merging**